### PR TITLE
2048+Chess: Replace DeprecatedString with String

### PR DESCRIPTION
**Pull request by aryanbaburajan**
Contributes to #17128.

First PR here, I'll be more than happy to correct any mistakes, if any. :)

### CODE REVIEW

File: Userland/Games/Chess/ChessWidget.cpp [hunk 1]
```diff
@@ -548,13 +547,13 @@ ErrorOr<void> ChessWidget::import_pgn(Core::File& file)
...
+    String movetext;
 
     for (size_t j = i; j < lines.size(); j++)
-        movetext = DeprecatedString::formatted("{}{}", movetext, lines.at(i).to_deprecated_string());
+        movetext = String::formatted("{}{}", movetext, lines.at(i)).release_value_but_fixme_should_propagate_errors();
 
-    for (auto token : movetext.split(' ')) {
-        token = token.trim_whitespace();
+    for (auto token : movetext.split(' ').release_value_but_fixme_should_propagate_errors()) {
+        token = token.trim_whitespace().release_value_but_fixme_should_propagate_errors();
```
Comment (Hendiadyoin1) [inline Userland/Games/Chess/ChessWidget.cpp:556]: You can safely use `TRY` here and down below

File: Userland/Games/2048/main.cpp
Comment (Hendiadyoin1) [inline Userland/Games/2048/main.cpp:1]: Can you squash that commit into the first one?
For that you use `git rebase`, see <https://youtu.be/ElRzTuYln0M>

**Review by Hendiadyoin1 — COMMENTED** [review]
Nice work!

File: Userland/Games/Chess/Engine.cpp [hunk 2]
```diff
@@ -43,9 +43,15 @@ void Engine::connect_to_engine_service()
     posix_spawn_file_actions_adddup2(&file_actions, wpipefds[0], STDIN_FILENO);
     posix_spawn_file_actions_adddup2(&file_actions, rpipefds[1], STDOUT_FILENO);
 
-    char const* argv[] = { m_command.characters(), nullptr };
+    auto command_length = m_command.code_points().length();
+    auto command_name = new char[command_length + 1];
```
Comment (Hendiadyoin1) [inline Userland/Games/Chess/Engine.cpp:47]: We seem to be leaking this,
can we safely `delete[]` this after `posix_spawnp`?

(We also have `Core::System::posix_spawnp` but that does not help much in this case)

File: Userland/Games/Chess/Engine.cpp [hunk 3]
```diff
@@ -18,7 +18,7 @@ Engine::~Engine()
     quit();
 }
 
-Engine::Engine(StringView command)
+Engine::Engine(String command)
     : m_command(command)
```
Comment (Hendiadyoin1) [inline Userland/Games/Chess/Engine.cpp:22]:
```suggestion
    : m_command(move(command))
```

**Review by Hendiadyoin1 — COMMENTED** [review]
A few more things after a more thorough and quite strict look

File: Userland/Games/Chess/Engine.cpp (hunk 2 above)
Comment (aryanbaburajan) [reply Userland/Games/Chess/Engine.cpp:47]: Freeing it after creating the new process (`posix_spawnp`) does seem to work.

File: Userland/Games/2048/main.cpp [hunk 4]
```diff
@@ -157,7 +158,7 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
         case Game::MoveOutcome::GameOver:
             update();
             GUI::MessageBox::show(window,
-                DeprecatedString::formatted("You reached {} in {} turns with a score of {}", game.largest_tile(), game.turns(), game.score()),
+                String::formatted("You reached {} in {} turns with a score of {}", game.largest_tile(), game.turns(), game.score()).release_value_but_fixme_should_propagate_errors(),
```
[the remark this answers is not in the document]
Comment (aryanbaburajan) [reply Userland/Games/2048/main.cpp:161]: Same situation as previous, the above code is in a lambda.

File: Userland/Games/2048/main.cpp [hunk 5]
```diff
@@ -143,7 +143,6 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
             break;
```
Comment (gmta) [inline Userland/Games/2048/main.cpp]: Commit message: "Fix: Fixes" does not help anyone; the commit title is enough for this change.

Comment (aryanbaburajan) [reply Userland/Games/2048/main.cpp]: Ah, apologies. Meant to remove that commit by squashing it through rebase but I guess it wasn't completely removed. Will fix it later today.

File: Userland/Games/Chess/ChessWidget.cpp [hunk 6]
```diff
@@ -648,13 +649,13 @@ ErrorOr<void> ChessWidget::export_pgn(Core::File& file) const
 
     // Movetext Section
     for (size_t i = 0, move_no = 1; i < m_board.moves().size(); i += 2, move_no++) {
-        const DeprecatedString white = m_board.moves().at(i).to_algebraic().release_value_but_fixme_should_propagate_errors().to_deprecated_string();
+        const String white = m_board.moves().at(i).to_algebraic().release_value_but_fixme_should_propagate_errors();
 
         if (i + 1 < m_board.moves().size()) {
-            const DeprecatedString black = m_board.moves().at(i + 1).to_algebraic().release_value_but_fixme_should_propagate_errors().to_deprecated_string();
-            TRY(file.write_until_depleted(DeprecatedString::formatted("{}. {} {} ", move_no, white, black).bytes()));
+            const String black = m_board.moves().at(i + 1).to_algebraic().release_value_but_fixme_should_propagate_errors();
```
Comment (LucasChollet) [inline Userland/Games/Chess/ChessWidget.cpp:655]: Can you move these `const`s east?

File: Userland/Games/Chess/ChessWidget.h [hunk 7]
```diff
@@ -140,7 +140,7 @@ class ChessWidget final
     Color m_marking_secondary_color { Color::from_argb(0x6655dd55) };
     Chess::Color m_side { Chess::Color::White };
     HashMap<Chess::Piece, RefPtr<Gfx::Bitmap const>> m_pieces;
-    DeprecatedString m_piece_set;
+    String m_piece_set;
```
Comment (LucasChollet) [inline Userland/Games/Chess/ChessWidget.h:143]: This value isn't used, so we can safely delete it

File: Userland/Games/Chess/ChessWidget.cpp [hunk 8]
```diff
@@ -522,9 +521,9 @@ void ChessWidget::playback_move(PlaybackDirection direction)
     update();
 }
 
-DeprecatedString ChessWidget::get_fen() const
+String ChessWidget::get_fen() const
```
Comment (LucasChollet) [inline Userland/Games/Chess/ChessWidget.cpp:524]: I'm not a fan of having a function returns a raw `String`, it's not better than a `DeprecatedString` IMO as someone will still need to make this a `ErrorOr<String>` later on. I would rather put the `release_value_but_fixme_should_propagate_errors()` on the caller side

File: Userland/Games/Chess/Engine.cpp [hunk 9]
```diff
@@ -43,13 +43,21 @@ void Engine::connect_to_engine_service()
     posix_spawn_file_actions_adddup2(&file_actions, wpipefds[0], STDIN_FILENO);
     posix_spawn_file_actions_adddup2(&file_actions, rpipefds[1], STDOUT_FILENO);
 
-    char const* argv[] = { m_command.characters(), nullptr };
+    auto command_length = m_command.code_points().length();
+    auto command_name = new char[command_length + 1];
+    memcpy(command_name, m_command.bytes_as_string_view().characters_without_null_termination(), command_length + 1);
```
Comment (gmta) [inline Userland/Games/Chess/Engine.cpp:48]: Isn't `command_length + 1` causing an out-of-bounds read here?

File: Userland/Games/2048/main.cpp [hunk 10]
```diff
@@ -143,7 +143,6 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
...
         case Game::MoveOutcome::Won: {
             update();
-
```
Comment (LucasChollet) [inline Userland/Games/2048/main.cpp]: Please avoid unrelated changes

File: Userland/Games/Chess/ChessWidget.cpp [hunk 11]
```diff
@@ -629,16 +627,18 @@ ErrorOr<void> ChessWidget::export_pgn(Core::File& file) const
     // Tag Pair Section
     TRY(file.write_until_depleted("[Event \"Casual Game\"]\n"sv.bytes()));
     TRY(file.write_until_depleted("[Site \"SerenityOS Chess\"]\n"sv.bytes()));
-    TRY(file.write_until_depleted(DeprecatedString::formatted("[Date \"{}\"]\n", Core::DateTime::now().to_deprecated_string("%Y.%m.%d"sv)).bytes()));
+    TRY(file.write_until_depleted(String::formatted("[Date \"{}\"]\n", Core::DateTime::now().to_deprecated_string("%Y.%m.%d"sv)).release_value_but_fixme_should_propagate_errors().bytes()));
```
Comment (LucasChollet) [inline Userland/Games/Chess/ChessWidget.cpp:630]: You can use `write_formatted()` here.
And that's a fallible context, so no need to `release_value_but_fixme_should_propagate_errors()`, you can use `TRY()`

File: Userland/Games/Chess/ChessWidget.cpp [hunk 12]
```diff
@@ -648,13 +648,13 @@ ErrorOr<void> ChessWidget::export_pgn(Core::File& file) const
 
     // Movetext Section
     for (size_t i = 0, move_no = 1; i < m_board.moves().size(); i += 2, move_no++) {
-        const DeprecatedString white = m_board.moves().at(i).to_algebraic().release_value_but_fixme_should_propagate_errors().to_deprecated_string();
+        String white = m_board.moves().at(i).to_algebraic().release_value_but_fixme_should_propagate_errors();
```
Comment (LucasChollet) [inline Userland/Games/Chess/ChessWidget.cpp:651]: `String const`, don't drop the `const`!
And again, you can try that

**Comment by aryanbaburajan** [comment]
Actually, there's one more bug to fix. No need for review just yet.

**Comment by aryanbaburajan** [comment]
Aforementioned bug should be fixed, please do review.

File: Userland/Games/2048/main.cpp (hunk 10 above)
Comment (linusg) [reply Userland/Games/2048/main.cpp]: This is not resolved

File: Userland/Games/Chess/ChessWidget.cpp [hunk 13]
```diff
@@ -629,16 +628,19 @@ ErrorOr<void> ChessWidget::export_pgn(Core::File& file) const
...
+    TRY(file.write_formatted("[Date \"{}\"]\n", Core::DateTime::now().to_deprecated_string("%Y.%m.%d"sv)));
     TRY(file.write_until_depleted("[Round \"1\"]\n"sv.bytes()));
 
-    DeprecatedString username(getlogin());
-    auto const player1 = (!username.is_empty() ? username.view() : "?"sv.bytes());
-    auto const player2 = (!m_engine.is_null() ? "SerenityOS ChessEngine"sv.bytes() : "?"sv.bytes());
-    TRY(file.write_until_depleted(DeprecatedString::formatted("[White \"{}\"]\n", m_side == Chess::Color::White ? player1 : player2).bytes()));
-    TRY(file.write_until_depleted(DeprecatedString::formatted("[Black \"{}\"]\n", m_side == Chess::Color::Black ? player1 : player2).bytes()));
+    auto current_user = TRY(Core::Account::self(Core::Account::Read::PasswdOnly));
+    String username = TRY(String::from_deprecated_string(current_user.username()));
```
Comment (linusg) [inline Userland/Games/Chess/ChessWidget.cpp:635]: `auto (const)`

File: Userland/Games/Chess/ChessWidget.cpp [hunk 14]
```diff
@@ -648,13 +650,13 @@ ErrorOr<void> ChessWidget::export_pgn(Core::File& file) const
 
     // Movetext Section
     for (size_t i = 0, move_no = 1; i < m_board.moves().size(); i += 2, move_no++) {
-        const DeprecatedString white = m_board.moves().at(i).to_algebraic().release_value_but_fixme_should_propagate_errors().to_deprecated_string();
+        String const white = m_board.moves().at(i).to_algebraic().release_value_but_fixme_should_propagate_errors();
```
Comment (linusg) [inline Userland/Games/Chess/ChessWidget.cpp:653]: `auto`

File: Userland/Games/Chess/ChessWidget.cpp [hunk 15]
```diff
@@ -648,13 +650,13 @@ ErrorOr<void> ChessWidget::export_pgn(Core::File& file) const
...
 
         if (i + 1 < m_board.moves().size()) {
-            const DeprecatedString black = m_board.moves().at(i + 1).to_algebraic().release_value_but_fixme_should_propagate_errors().to_deprecated_string();
-            TRY(file.write_until_depleted(DeprecatedString::formatted("{}. {} {} ", move_no, white, black).bytes()));
+            String black = m_board.moves().at(i + 1).to_algebraic().release_value_but_fixme_should_propagate_errors();
```
Comment (linusg) [inline Userland/Games/Chess/ChessWidget.cpp:656]: `auto`

File: AK/String.cpp [hunk 16]
```diff
@@ -484,6 +484,11 @@ ErrorOr<String> String::trim(StringView code_points_to_trim, TrimMode mode) cons
     return trim(Utf8View { code_points_to_trim }, mode);
 }
 
+ErrorOr<String> String::trim_whitespace(TrimMode mode) const
+{
+    return trim(" \n\t\v\f\r"sv, mode);
```
Comment (linusg) [inline AK/String.cpp:489]: Let's not repeat mistakes from the old string type, "whitespace" has no unique definition. At the very least call it `trim_ascii_whitespace`.

**Review by linusg — CHANGES_REQUESTED** [review]

File: Userland/Games/Chess/main.cpp [hunk 17]
```diff
@@ -69,6 +69,7 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
     for (auto const& engine : engines)
         TRY(Core::System::unveil(engine.path, "x"sv));
 
+    TRY(Core::System::unveil("/etc", "r"));
```
[the remark this answers is not in the document]
Comment (aryanbaburajan) [reply Userland/Games/Chess/main.cpp:72]: It's required for `Core::Account::self()`, otherwise it fails to access `/etc/passwd` as it has not been unveiled.

File: AK/String.cpp [hunk 18]
```diff
@@ -484,7 +484,7 @@ ErrorOr<String> String::trim(StringView code_points_to_trim, TrimMode mode) cons
     return trim(Utf8View { code_points_to_trim }, mode);
 }
 
-ErrorOr<String> String::trim_whitespace(TrimMode mode) const
+ErrorOr<String> String::trim_ascii_whitespace(TrimMode mode) const
```
Comment (LucasChollet) [inline AK/String.cpp:487]: Don't rename it in the Chess commit, add the method with the correct name int the first place

File: Userland/Games/Chess/main.cpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (LucasChollet) [reply Userland/Games/Chess/main.cpp:72]: You can just unveil "/etc/passwd" then, and maybe add a comment to explain that. Unveiling this file in a Chess game looks really suspicious.

**Review by gmta — APPROVED** [review]

**Comment by aryanbaburajan** [comment]
Oh, thank you. I was busy with schoolwork.
